### PR TITLE
feat(lsp): basic support for textDocument/completion

### PR DESCRIPTION
**Pull request by lucacasonato**
This adds basic `textDocument/completion` handling.

Currently text completions with `TextEdit`s do not work. I have not managed to figure out why though. Until I find the cause I have disabled the `includeCompletionsWithInsertText` TS user preference which should mitigate this.

```ts
function str(s?: Deno.File) {
  s.
//  ^- complete here is not as nice as VSCode built in. (optional chaining)
}
```

In the future we should implement `completionItem/resolve` to support `documentation` and `details` field in the `CompletionItem`.

### CODE REVIEW

File: cli/lsp/tsc.rs [hunk 1]
```diff
@@ -519,6 +559,111 @@ impl ReferenceEntry {
...
+        item.insert_text_format = Some(lsp_types::InsertTextFormat::Snippet);
+      }
+      _ => {}
+    }
+
+    let mut insert_text = self.insert_text;
+    let replacement_range: Option<lsp_types::Range> =
+      self.replacement_span.map(|span| span.to_range(line_index));
+
+    // // Make sure we only replace a single line at most
```
Comment (kitsonk) [inline cli/lsp/tsc.rs:633]: Probably should clean this up, make a TODO, or something...

File: cli/tsc/99_main_compiler.js [hunk 2]
```diff
@@ -471,6 +471,10 @@ delete Object.prototype.__proto__;
     core.jsonOpSync("op_respond", { id, data });
   }
 
+  const userPreferences = {
```
Comment (kitsonk) [inline cli/tsc/99_main_compiler.js:474]:
```suggestion
  /** @type {ts.UserPreferences} */
  const userPreferences = {
```

I think it is also debatable if this were moved to Rust as well, as sort of want zero config in JavaScript if we can avoid it.

Comment (kitsonk) [reply cli/tsc/99_main_compiler.js:474]: As we will likely need to change/update/manage these on the lsp side, especially when we need to do refactorings... We could do it like the configuration managements as well...

```ts
export interface UserPreferences {
        readonly disableSuggestions?: boolean;
        readonly quotePreference?: "auto" | "double" | "single";
        readonly includeCompletionsForModuleExports?: boolean;
        readonly includeAutomaticOptionalChainCompletions?: boolean;
        readonly includeCompletionsWithInsertText?: boolean;
        readonly importModuleSpecifierPreference?: "auto" | "relative" | "non-relative";
        /** Determines whether we import `foo/index.ts` as "foo", "foo/index", or "foo/index.js" */
        readonly importModuleSpecifierEnding?: "auto" | "minimal" | "index" | "js";
        readonly allowTextChangesInNewFiles?: boolean;
        readonly providePrefixAndSuffixTextForRename?: boolean;
        readonly includePackageJsonAutoImports?: "auto" | "on" | "off";
        readonly provideRefactorNotApplicableReason?: boolean;
    }
```

**Review by kitsonk — COMMENTED** [review]
a couple thoughts...

File: cli/tsc/99_main_compiler.js (hunk 2 above)
Comment (lucacasonato) [reply cli/tsc/99_main_compiler.js:474]: Good point. Let me move it.

**Review by kitsonk — APPROVED** [review]
LGTM 🥳

**Pull request merged**